In [1]:
import pandas as pd
import numpy as np

In [14]:
data='D:/MAchine Learning/Dataset/Sept-March data - Sheet.csv'
df=pd.read_csv(data,sep=',')

Problem Statement:
1. Classifcation Problem: To predict if the ticket's SLA would get violated or not.
2. Regression Probelm: To predict time required to resolve a support ticket(continued in 2nd file).

In [ ]:
Following code includes: Pre-processing
                         Model Building
                         HTML code genrator
                         Pickle file
P.s. : EDA removed

In [1]:
#df.head()

Pre-Processing

In [16]:
df.drop(['Ticket ID', 'Subject', 'Status', 'Source','Created time', 'Due by Time', 'Resolved time', 'Closed time',
       'Last update time', 'Initial response time', 'Time tracked',
       'First response time (in hrs)', 'Resolution time (in hrs)',
       'Agent interactions', 'Customer interactions', 'First response status', 'Tags', 'Survey results', 'Association type',
       'Internal agent', 'Internal group', 'Every response status', 
        'Biz ID', 'Biz Name', 'POS',
       'POS Name', 'Aggregator name', 'Source.1',
       'Order Status', 'Username', 'UrbanPiper Store Name/ID', 'Item Name/ID',
       'Product Name', 'Type of issue', 'Order ID', 'Username.1', 'Country',
       'Option Name/ID', 'Email Address', 'Date & Time', 'Full name',
       'Contact ID'],axis=1,inplace=True)

In [17]:
df.drop(['Product'],axis=1,inplace=True)

In [18]:
df["Product.1"] =df["Product.1"].replace({np.nan:'Others'}, regex=True)
df["Merchant"] =df["Merchant"].replace({np.nan:'Others'}, regex=False)


In [19]:
df=df.dropna() #Dropping all cases without Resolution status

In [20]:
df.nunique() # Unique Values

Priority                   4
Type                       7
Agent                     14
Group                      5
Resolution status          2
Product.1                  6
Module                     6
Merchant                 378
Resolution Dependency      9
dtype: int64

In [21]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

In [ ]:
#Converting Categorical data to encoded Data

In [22]:
df['Type']=le.fit_transform(df['Type']) 

C:\Users\lenov\AppData\Local\Temp/ipykernel_9300/2525559946.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Type']=le.fit_transform(df['Type']) #Converting Categorical data to encoded Data


In [233]:
#Creating Data Frame having Categorical Data and Unique Encoded Value
e_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Type']) 



In [234]:
le.classes_ #No of unique Classes

array(['Demo request', 'FYI', 'Feature request', 'Issues',
       'Misconfiguration', 'Others', 'Service request'], dtype=object)

Since we would need options for a particular category, manually genrating the required HTML format could be difficult, especially for large number of options, Hence the code below automates the work for us

In [236]:
# Converting DF to required HTML format
type=[]
for i in range(len(e_type)):
    type.append('''<option value="{}">{}</option>'''.format(i,e_type.loc[i,'Type']))
for i in type:
    print(i)

<option value="0">Demo request</option>
<option value="1">FYI</option>
<option value="2">Feature request</option>
<option value="3">Issues</option>
<option value="4">Misconfiguration</option>
<option value="5">Others</option>
<option value="6">Service request</option>


In [261]:
#Repeating each transformation seperately as fit function is changing for every row

In [255]:
df['Priority']=le.fit_transform(df['Priority'])
p_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Priority'])
p=[]
for i in range(len(p_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,p_type.loc[i,'Priority']))
for i in p:
    print(i)

<option value="0">High</option>
<option value="1">Low</option>
<option value="2">Medium</option>
<option value="3">Urgent</option>


In [ ]:
df['Agent']=le.fit_transform(df['Agent'])
a_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Agent'])
p=[]
for i in range(len(a_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,a_type.loc[i,'Agent']))
for i in p:
    print(i)

In [257]:
df['Group']=le.fit_transform(df['Group'])
g_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Group'])
p=[]
for i in range(len(g_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,g_type.loc[i,'Group']))
for i in p:
    print(i)


<option value="0">Call Support Team</option>
<option value="1">Customer Experience</option>
<option value="2">Firefighters—L1</option>
<option value="3">Firefighters—L2</option>
<option value="4">No Group</option>


In [258]:
df['Product.1']=le.fit_transform(df['Product.1'])
g_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Product.1'])
p=[]
for i in range(len(g_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,g_type.loc[i,'Product.1']))
for i in p:
    print(i)

<option value="0">Comet</option>
<option value="1">Hub</option>
<option value="2">Meraki</option>
<option value="3">Others</option>
<option value="4">Prime</option>
<option value="5">Prime POS</option>


In [259]:
df['Module']=le.fit_transform(df['Module'])
g_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Product.1'])
p=[]
for i in range(len(g_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,g_type.loc[i,'Product.1']))
for i in p:
    print(i)


<option value="0">Catalogue Config Request</option>
<option value="1">Charges</option>
<option value="2">Demo Request</option>
<option value="3">Store Availability</option>
<option value="4">Taxes</option>
<option value="5">User Account</option>


In [ ]:
df['Merchant']=le.fit_transform(df['Merchant'])
g_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Product.1'])
p=[]
for i in range(len(g_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,g_type.loc[i,'Product.1']))
for i in p:
    print(i)

In [262]:
df['Resolution Dependency']=le.fit_transform(df['Resolution Dependency'])
g_type=pd.DataFrame(le.classes_,le.transform(le.classes_),columns=['Product.1'])
p=[]
for i in range(len(g_type)):
    p.append('''<option value="{}">{}</option>'''.format(i,g_type.loc[i,'Product.1']))
for i in p:
    print(i)


<option value="0">Account Management</option>
<option value="1">Finance</option>
<option value="2">Integrations</option>
<option value="3">L1 Support</option>
<option value="4">L2 Support</option>
<option value="5">On-boarding</option>
<option value="6">Product Specialist</option>
<option value="7">Program Management</option>
<option value="8">Sales</option>


In [ ]:
#Declaring Dependant & Independant variable 

In [45]:
X=df.drop(['Resolution status'],axis=1)
y=df['Resolution status']

In [263]:
y.value_counts()

Within SLA      2467
SLA Violated     631
Name: Resolution status, dtype: int64

In [264]:
#As we can see our target variable is Imbalanced, Balancing the same by doing a OverSampling

In [266]:
from imblearn.over_sampling import SMOTE
sm=SMOTE()

In [267]:
X_os,y_os=sm.fit_resample(X,y)

In [268]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_os,y_os,test_size=0.3,random_state=123)

In [269]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [270]:
from sklearn.tree import DecisionTreeClassifier

In [271]:
dt=DecisionTreeClassifier()

In [272]:
dt.fit(X_train,y_train)

DecisionTreeClassifier()

In [273]:
y_pred=dt.predict(X_test)


In [274]:
accuracy_score(y_test,y_pred)

0.7981093855503039

#Observations: Since the no of merchants in the independant vaeriable are close 378 to  as compared to other independant feature which range between 0 to 30, its understood that merchant feature needs to be normailze
                                 But as opposed to that Decison Tree is performing better with an accuracy of 79% with non normalized data as comapred to 76% when data is normalized.
                             


In [275]:
import pickle

In [276]:
#Creating Pickel Model

pickle.dump(dt, open('model_Pickle_dump.pkl','wb'))

# Loading model to compare the results
model = pickle.load(open('model_Pickle_dump.pkl','rb'))
print(model.predict([[2, 9, 6, 4 ,5 ,7 ,8 ,9]]))


['SLA Violated']


C:\Users\lenov\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(
